File: node/AssemblerWeb/appSitesConfig.js
import fs from 'fs';
import path from 'path';

/**
 * Discovers AppSites by scanning the AppSites folder and generates appsites.csv
 */
function generateAppSitesCsv(wwwrootPath) {
  const appSitesPath = path.join(wwwrootPath, 'AppSites');
  const appDataPath = path.join(wwwrootPath, 'App_Data');
  const csvFilePath = path.join(appDataPath, 'appsites.csv');

  if (!fs.existsSync(appSitesPath)) {
    throw new Error(`AppSites directory not found: ${appSitesPath}`);
  }

  // Ensure App_Data directory exists
  if (!fs.existsSync(appDataPath)) {
    fs.mkdirSync(appDataPath, { recursive: true });
  }

  // Get all directories in AppSites folder
  const entries = fs.readdirSync(appSitesPath, { withFileTypes: true });
  let appSites = entries
    .filter(entry => entry.isDirectory())
    .map(entry => entry.name);

  // Add Index as it's a valid AppSite
  if (!appSites.some(s => s.toLowerCase() === 'index')) {
    appSites.push('Index');
  }

  // Sort for consistency
  appSites.sort();

  // Write as CSV (comma-delimited)
  const csv = appSites.join(',');
  fs.writeFileSync(csvFilePath, csv);

  console.log(`[AppSitesConfig] Generated appsites.csv with ${appSites.length} AppSites`);
}

/**
 * Loads AppSites from appsites.csv, generates it if it doesn't exist
 */
export function loadAppSites(wwwrootPath) {
  const appDataPath = path.join(wwwrootPath, 'App_Data');
  const csvFilePath = path.join(appDataPath, 'appsites.csv');

  // Generate appsites.csv if it doesn't exist
  if (!fs.existsSync(csvFilePath)) {
    console.log('[AppSitesConfig] appsites.csv not found, generating...');
    generateAppSitesCsv(wwwrootPath);
  }

  // Read and parse CSV
  const csv = fs.readFileSync(csvFilePath, 'utf8').trim();

  if (!csv) {
    throw new Error('appsites.csv is empty');
  }

  const appSites = csv
    .split(',')
    .map(s => s.trim())
    .filter(s => s.length > 0);

  if (appSites.length === 0) {
    throw new Error('No AppSites found in appsites.csv');
  }

  console.log(`[AppSitesConfig] Loaded ${appSites.length} AppSites from appsites.csv`);

  // Return as Set for case-insensitive comparison
  return new Set(appSites.map(s => s.toLowerCase()));
}

/**
 * Reloads AppSites by regenerating appsites.csv from the file system
 */
export function reloadAppSites(wwwrootPath) {
  console.log('[AppSitesConfig] Reloading AppSites...');
  generateAppSitesCsv(wwwrootPath);
  return loadAppSites(wwwrootPath);
}

File: rust/AssemblerWeb/src/app_sites_config.rs
use std::collections::HashSet;
use std::fs;
use std::path::Path;

/// Discovers AppSites by scanning the AppSites folder and generates appsites.csv
fn generate_app_sites_csv(wwwroot_path: &str) -> Result<(), String> {
    let app_sites_path = Path::new(wwwroot_path).join("AppSites");
    let app_data_path = Path::new(wwwroot_path).join("App_Data");
    let csv_file_path = app_data_path.join("appsites.csv");

    if !app_sites_path.exists() {
        return Err(format!("AppSites directory not found: {:?}", app_sites_path));
    }

    // Ensure App_Data directory exists
    if !app_data_path.exists() {
        fs::create_dir_all(&app_data_path)
            .map_err(|e| format!("Failed to create App_Data directory: {}", e))?;
    }

    // Get all directories in AppSites folder
    let mut app_sites: Vec<String> = fs::read_dir(&app_sites_path)
        .map_err(|e| format!("Failed to read AppSites directory: {}", e))?
        .filter_map(|entry| entry.ok())
        .filter(|entry| entry.path().is_dir())
        .filter_map(|entry| {
            entry.file_name().to_str().map(|s| s.to_string())
        })
        .collect();

    // Add Index as it's a valid AppSite
    if !app_sites.iter().any(|s| s.eq_ignore_ascii_case("Index")) {
        app_sites.push("Index".to_string());
    }

    // Sort for consistency
    app_sites.sort();

    // Write as CSV (comma-delimited)
    let csv = app_sites.join(",");
    fs::write(&csv_file_path, csv)
        .map_err(|e| format!("Failed to write appsites.csv: {}", e))?;

    println!("[AppSitesConfig] Generated appsites.csv with {} AppSites", app_sites.len());
    Ok(())
}

/// Loads AppSites from appsites.csv, generates it if it doesn't exist
pub fn load_app_sites(wwwroot_path: &str) -> Result<HashSet<String>, String> {
    let app_data_path = Path::new(wwwroot_path).join("App_Data");
    let csv_file_path = app_data_path.join("appsites.csv");

    // Generate appsites.csv if it doesn't exist
    if !csv_file_path.exists() {
        println!("[AppSitesConfig] appsites.csv not found, generating...");
        generate_app_sites_csv(wwwroot_path)?;
    }

    // Read and parse CSV
    let csv = fs::read_to_string(&csv_file_path)
        .map_err(|e| format!("Failed to read appsites.csv: {}", e))?;

    let csv = csv.trim();
    if csv.is_empty() {
        return Err("appsites.csv is empty".to_string());
    }

    let app_sites: Vec<String> = csv.split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    if app_sites.is_empty() {
        return Err("No AppSites found in appsites.csv".to_string());
    }

    println!("[AppSitesConfig] Loaded {} AppSites from appsites.csv", app_sites.len());

    // Use case-insensitive comparison
    Ok(app_sites.into_iter().collect())
}

/// Reloads AppSites by regenerating appsites.csv from the file system
#[allow(dead_code)]
pub fn reload_app_sites(wwwroot_path: &str) -> Result<HashSet<String>, String> {
    println!("[AppSitesConfig] Reloading AppSites...");
    generate_app_sites_csv(wwwroot_path)?;
    load_app_sites(wwwroot_path)
}
